### PR TITLE
Refactor JDBC URL parsing

### PR DESCRIPTION
**Pull request by rnorth**
Credit belongs mainly to @justinwyer. Split out from #4382 

- Detect SID `:` or service name `/` in the supplied TC jdbc URL and honor this in the underlying JDBC URL.  
- Switched affected regex patterns to use named groups  
- Add tests for both SID and service name connections
- Further tidying
- Fix image tag extraction Tags containing any of the chars [t,h,i,n] would break

### CODE REVIEW

**Comment by kiview** [comment]
japicmp is faling on CI, but I don't see the change that leads to this. Can't check right now locally, because japicmp does not run on our Windows build.

File: modules/jdbc/src/main/java/org/testcontainers/jdbc/ConnectionUrl.java [hunk 1]
```diff
@@ -209,26 +209,52 @@ private void parseUrl() {
...
+                "@" +
+                "(?<dbHostString>[^?]+)" +
+                "(?<queryParameters>\\?.*)?"
+        );
 
         //Matches to part of string - hostname:port/databasename
-        Pattern DB_INSTANCE_MATCHING_PATTERN = Pattern.compile("([^:]+)(:([0-9]+))?/([^\\\\?]+)");
-
-        Pattern DAEMON_MATCHING_PATTERN = Pattern.compile(".*([\\?&]?)TC_DAEMON=([^\\?&]+).*");
-        Pattern INITSCRIPT_MATCHING_PATTERN = Pattern.compile(".*([\\?&]?)TC_INITSCRIPT=([^\\?&]+).*");
```
Comment (kiview) [inline modules/jdbc/src/main/java/org/testcontainers/jdbc/ConnectionUrl.java:220]: japicmp failed because of this removed `Pattern`.

**Comment by rnorth** [comment]
Ergh; I'll fix - I removed an unused constant without realising that it's public 😞.

**Review by kiview — APPROVED** [review]
LGTM assuming CI is now green 🙂

**Pull request merged**